plugin/core/main.py

```diff
@@ -1,29 +1,131 @@
+from .handlers import LanguageHandler
 from .logging import set_debug_logging, set_exception_logging
 from .panels import destroy_output_panels, ensure_panel, PanelName
 from .popups import popups
-from .registry import windows, load_handlers, unload_sessions
+from .protocol import Response
+from .protocol import WorkspaceFolder
+from .registry import windows, unload_sessions
+from .rpc import method2attr
+from .sessions import AbstractPlugin
+from .sessions import register_plugin
+from .sessions import Session
 from .settings import settings, load_settings, unload_settings
-from .typing import Optional
+from .transports import kill_all_subprocesses
+from .types import ClientConfig
+from .typing import Optional, List, Type, Callable, Dict
+import weakref
 
 
 def ensure_server_panel(window: sublime.Window) -> Optional[sublime.View]:
     return ensure_panel(window, PanelName.LanguageServers, "", "", "Packages/LSP/Syntaxes/ServerLog.sublime-syntax")
 
 
+def _get_final_subclasses(derived: List[Type], results: List[Type]) -> None:
+    """
+    This function should be removed: https://github.com/sublimelsp/LSP/issues/899
+    """
+    for d in derived:
+        d_subclasses = d.__subclasses__()
+        if len(d_subclasses) > 0:
+            _get_final_subclasses(d_subclasses, results)
+        else:
+            results.append(d)
+
+
+def _forcefully_register_plugins() -> None:
+    """
+    This function should be removed: https://github.com/sublimelsp/LSP/issues/899
+    """
+    plugin_classes = []  # type: List[Type[AbstractPlugin]]
+    _get_final_subclasses(AbstractPlugin.__subclasses__(), plugin_classes)
+    for plugin_class in plugin_classes:
+        register_plugin(plugin_class)
+    language_handler_classes = []  # type: List[Type[LanguageHandler]]
+    _get_final_subclasses(LanguageHandler.__subclasses__(), language_handler_classes)
+    for language_handler_class in language_handler_classes:
+        # Create an ephemeral plugin that stores an instance of the LanguageHandler as a class instance. Custom requests
+        # and notifications will work.
+        class LanguageHandlerTransition(AbstractPlugin):
+
+            handler = language_handler_class()
+            has_regular_name = False
+
+            @classmethod
+            def name(cls) -> str:
+                n = cls.handler.name  # type: ignore
+                if n.lower().startswith('lsp-'):
+                    cls.has_regular_name = True
+                    return n[len('lsp-'):]
+                return n
+
+            @classmethod
+            def configuration(cls) -> sublime.Settings:
+                if cls.has_regular_name:
+                    settings_filename = "LSP-{}.sublime-settings".format(cls.name())
+                else:
+                    # e.g. "metals-sublime.sublime-settings"
+                    settings_filename = "{}.sublime-settings".format(cls.name())
+                settings = sublime.load_settings(settings_filename)
```

Comment:
I think now that returning a `sublime.Settings` object might not be a good choice. The problem if this settings object is created with the `sublime.load_settings()` function, is, that keys from the user settings completely overwrite those which are predefined in the packages' settings file. This is especially problematic for the `"settings": {}` dict, if I want to only change a single value, without loosing all other values. Therefore we should recursively merge dicts from the default and user settings instead.

I adopted the code from LSP-lemminx [here](https://github.com/sublimelsp/LSP-lemminx/blob/master/plugin.py#L24-L47) and changed some details (for example I think lists should not be merged, since it could lead to problems with the "command" setting).

```python
import sublime


def merge_dicts(default: dict, user: dict):
    """ Recursively merges user into default """
    for key, value in default.items():
        if key in user:
            new_value = user[key]
            if isinstance(value, dict) and isinstance(new_value, dict):
                yield key, dict(merge_dicts(value, new_value))
            # elif isinstance(value, list) and isinstance(new_value, list):
            #     yield key, list(set(value) | set(new_value))
            else:
                # overwrite default value with user value
                yield key, new_value
        else:
            yield key, value
    # add all additional items from user
    for key, value in user.items():
        if key not in default:
            yield key, value


def load_settings(base_name: str) -> dict:
    default_settings_json = sublime.load_resource("Packages/{}/{}".format(__package__, base_name))
    user_settings_json = sublime.load_resource("Packages/User/{}".format(base_name))
    default_settings = sublime.decode_value(default_settings_json)
    user_settings = sublime.decode_value(user_settings_json)
    settings = {"enabled": True}
    for key, value in merge_dicts(default_settings, user_settings):
        settings[key] = value
    return settings



# LanguageHandler
def config(self):
    settings = load_settings("LSP-julia.sublime-settings")
    return read_client_config(self.name, settings)
```

We should add something like this here or put it into the lsp_utils, so that LSP-* packages don't have to reinvent the wheel everytime.

Reply:
Well, one important property of sublime.Settings objects is that you can get notifications about them when they change. That's an important thing to have to be able to get to a state of this package where we can do hot-reloading when some setting somewhere changed. There's been a lot of users that get confused because they've changed some setting and nothing happens. Only when restarting ST.

Calling `sublime.load_resource` and `sublime.decode_value` is also a waste of CPU cycles because the settings are already in memory.

However I do understand the concern and I suppose we would have to have this merge_dicts function for a while.

The discussion in #1048 should lead to something better where we don't use `sublime.load_resource` / `sublime.decode_value`. I still say a separate `server-settings.sublime-settings` file in Packages/User works best, because:

1) We read the (dotted, since ST4) key-value pairs from `"settings"` from the sublime.Settings object backed by some LSP-foobar.sublime-settings file.
2) For each of those key-value pairs we do `server_settings.get(key, default=value)` where `server_settings` is also a sublime.Settings object but backed by Packages/User/server-settings.sublime-settings. This way the `"settings"` from LSP-foobar.sublime-settings files play the role of sane defaults and also provide the keys to look up in server-settings.sublime-settings. Hope this all makes sense.

Reply:
Hmm I realize now that we would still have to use `sublime.load_resource("Packages/User/server-settings.sublime-settings")` because of dotted sections like `cpp.formatterOptions`. We can't do `settings.get("cpp.formatterOptions")` and expect it to match `cpp.formatterOptions.foo` and `cpp.formatterOptions.bar`.

Reply:
Let's keep the api as-is and fix outstanding issues in another PR.

Now that dotted-keys notation works in ST4 I'm fairly confident that a separate `Packages/User/server-settings.sublime-settings` for all user overrides will work fine.